### PR TITLE
Fix visible data clobbering by progress bar in client in interactive mode

### LINKED ISSUE #12562: JSON and JSONCompact formats produce invalid JSON objects in interactive mode of clickhouse-client
The last line in `data` section of the output is always empty. 

```
# clickhouse-client
ClickHouse client version 20.5.2.7 (official build).
Connecting to localhost:9001 as user default.
Connected to ClickHouse server version 20.5.2 revision 54435.

sas-vrxctd96rqw5lw25.db.yandex.net :) select 1 format JSON

SELECT 1
FORMAT JSON

{
	"meta":
	[
		{
			"name": "1",
			"type": "UInt8"
		}
	],

	"data":
	[
		{
			"1": 1

	],

	"rows": 1,

	"statistics":
	{
		"elapsed": 0.00019223,
		"rows_read": 1,
		"bytes_read": 1
	}
}

```

The issue appears only in interactive mode. In non-interactive mode clickhouse-client produces the correct output:

```
# clickhouse-client -q 'select 1 format JSON'
{
	"meta":
	[
		{
			"name": "1",
			"type": "UInt8"
		}
	],

	"data":
	[
		{
			"1": 1
		}
	],

	"rows": 1,

	"statistics":
	{
		"elapsed": 0.000220845,
		"rows_read": 1,
		"bytes_read": 1
	}
}
```

### LINKED ISSUE #12964: Progress bar flicker in clickhouse-client.
When I execute the following query:

```
SELECT 
    toInt64(WatchID), 
    toInt8(JavaEnable), 
    toValidUTF8(toString(Title)), 
    toInt16(GoodEvent), 
    EventTime, 
    EventDate, 
    toInt32(CounterID), 
    toInt32(ClientIP), 
    toInt32(RegionID), 
    toInt64(UserID), 
    toInt8(CounterClass), 
    toInt8(OS), 
    toInt8(UserAgent), 
    toValidUTF8(toString(URL)), 
    toValidUTF8(toString(Referer)), 
    toInt8(Refresh), 
    toInt16(RefererCategoryID), 
    toInt32(RefererRegionID), 
    toInt16(URLCategoryID), 
    toInt32(URLRegionID), 
    toInt16(ResolutionWidth), 
    toInt16(ResolutionHeight), 
    toInt8(ResolutionDepth), 
    toInt8(FlashMajor), 
    toInt8(FlashMinor), 
    toValidUTF8(toString(FlashMinor2)), 
    toInt8(NetMajor), 
    toInt8(NetMinor), 
    toInt16(UserAgentMajor), 
    toValidUTF8(toString(UserAgentMinor)), 
    toInt8(CookieEnable), 
    toInt8(JavascriptEnable), 
    toInt8(IsMobile), 
    toInt8(MobilePhone), 
    toValidUTF8(toString(MobilePhoneModel)), 
    toValidUTF8(toString(Params)), 
    toInt32(IPNetworkID), 
    toInt8(TraficSourceID), 
    toInt16(SearchEngineID), 
    toValidUTF8(toString(SearchPhrase)), 
    toInt8(AdvEngineID), 
    toInt8(IsArtifical), 
    toInt16(WindowClientWidth), 
    toInt16(WindowClientHeight), 
    toInt16(ClientTimeZone), 
    ClientEventTime, 
    toInt8(SilverlightVersion1), 
    toInt8(SilverlightVersion2), 
    toInt32(SilverlightVersion3), 
    toInt16(SilverlightVersion4), 
    toValidUTF8(toString(PageCharset)), 
    toInt32(CodeVersion), 
    toInt8(IsLink), 
    toInt8(IsDownload), 
    toInt8(IsNotBounce), 
    toInt64(FUniqID), 
    toValidUTF8(toString(OriginalURL)), 
    toInt32(HID), 
    toInt8(IsOldCounter), 
    toInt8(IsEvent), 
    toInt8(IsParameter), 
    toInt8(DontCountHits), 
    toInt8(WithHash), 
    toValidUTF8(toString(HitColor)), 
    LocalEventTime, 
    toInt8(Age), 
    toInt8(Sex), 
    toInt8(Income), 
    toInt16(Interests), 
    toInt8(Robotness), 
    toInt32(RemoteIP), 
    toInt32(WindowName), 
    toInt32(OpenerName), 
    toInt16(HistoryLength), 
    toValidUTF8(toString(BrowserLanguage)), 
    toValidUTF8(toString(BrowserCountry)), 
    toValidUTF8(toString(SocialNetwork)), 
    toValidUTF8(toString(SocialAction)), 
    toInt16(HTTPError), 
    toInt32(SendTiming), 
    toInt32(DNSTiming), 
    toInt32(ConnectTiming), 
    toInt32(ResponseStartTiming), 
    toInt32(ResponseEndTiming), 
    toInt32(FetchTiming), 
    toInt8(SocialSourceNetworkID), 
    toValidUTF8(toString(SocialSourcePage)), 
    toInt64(ParamPrice), 
    toValidUTF8(toString(ParamOrderID)), 
    toValidUTF8(toString(ParamCurrency)), 
    toInt16(ParamCurrencyID), 
    toValidUTF8(toString(OpenstatServiceName)), 
    toValidUTF8(toString(OpenstatCampaignID)), 
    toValidUTF8(toString(OpenstatAdID)), 
    toValidUTF8(toString(OpenstatSourceID)), 
    toValidUTF8(toString(UTMSource)), 
    toValidUTF8(toString(UTMMedium)), 
    toValidUTF8(toString(UTMCampaign)), 
    toValidUTF8(toString(UTMContent)), 
    toValidUTF8(toString(UTMTerm)), 
    toValidUTF8(toString(FromTag)), 
    toInt8(HasGCLID), 
    toInt64(RefererHash), 
    toInt64(URLHash), 
    toInt32(CLID)
FROM hits_100m_obfuscated
INTO OUTFILE '/home/milovidov/example_datasets/hits_100m_obfuscated.csv'
```

### LINKED ISSUE #13369: Invalid json output with FORMAT JSON
Easy to reproduce:
### 20.3.13
```
SELECT 1
FORMAT JSON

{
        "meta":
        [
                {
                        "name": "1",
                        "type": "UInt8"
                }
        ],

        "data":
        [
                {
                        "1": 1
                }
        ],

        "rows": 1,

        "statistics":
        {
                "elapsed": 0.000203731,
                "rows_read": 1,
                "bytes_read": 1
        }
}
```

### 20.4.5
```
:) select 1 format JSON

SELECT 1
FORMAT JSON

{
        "meta":
        [
                {
                        "name": "1",
                        "type": "UInt8"
                }
        ],

        "data":
        [
                {
                        "1": 1
>>>>>>>> no closing bracket <<<<<<<<<<
        ],

        "rows": 1,

        "statistics":
        {
                "elapsed": 0.000771732,
                "rows_read": 1,
                "bytes_read": 1
        }
}
```
### master
```

SELECT 1
FORMAT JSON

{
        "meta":
        [
                {
                        "name": "1",
                        "type": "UInt8"
                }
        ],

        "data":
        [
                {
                        "1": 1
>>>>>>>> no closing bracket <<<<<<<<<<
        ],

        "rows": 1,

        "statistics":
        {
                "elapsed": 0.00103364,
                "rows_read": 1,
                "bytes_read": 1
        }
}

```

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix visible data clobbering by progress bar in client in interactive mode. This fixes #12562 and #13369 and #13584 and fixes #12964.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
Test 1:
`select 1 format JSON`
should output valid JSON with no progress bar in the middle.

Test 2:
`select count() FROM system.numbers format JSON`
while query is run, progress indicator is displayed and updated;
after you press Ctrl+C it should output valid JSON with no progress bar in the middle.

Test 3:
```
SET max_block_size = 1;
select sleep(1) FROM system.numbers format Values
```
should output correct Values format and progress bar.
Also the same with Pretty format.

Test 4:
```
SET max_block_size = 1;
select sleep(1) FROM system.numbers format Values
```
Progress bar should not flicker.

Test 5:
```
SET max_block_size = 1
SELECT * FROM system.numbers INTO OUTFILE 'test.tsv'
```
Progress bar should not flicker.

Test 6:
```
SELECT * FROM system.numbers
```
Progress bar should not flicker. *But it still flickers after output of 10 000 rows!*

**Comment by alexey-milovidov** [comment]
It's very difficult to automate these test.

For example, to check that the text displayed in a terminal represents valid JSON, we have to use terminal emulator.

**Comment by KochetovNicolai** [comment]
Tested between two versions.

#### Test 1. 
JSON is valid now, `{` is here.
> no progress bar in the middle

It works in the same (correct) way for both versions.

#### Test 2. 
Old:
```
Cancelling query.
↑ Progress: 2.60 billion rows, 20.81 GB (1.85 billion rows/s., 14.81 GB/s.) {
	"meta":
	[
...
Ok.
Query was cancelled.

0 rows in set. Elapsed: 1.405 sec. Processed 2.60 billion rows, 20.81 GB (1.85 billion rows/s., 14.81 GB/s.) 
```
New:
```
Cancelling query.
{
	"meta":
	[
...
Ok.
Query was cancelled.

0 rows in set. Elapsed: 1.204 sec. Processed 2.29 billion rows, 18.30 GB (1.90 billion rows/s., 15.19 GB/s.) 

```

Fixed.

#### Test 3.
Old:
```
select sleep(1) FROM system.numbers format Values

SELECT sleep(1)
FROM system.numbers
FORMAT Values

Cancelling query.
Query was cancelled.

2 rows in set. Elapsed: 3.003 sec. 
```
New:
```
:) select sleep(1) FROM system.numbers format Values

SELECT sleep(1)
FROM system.numbers
FORMAT Values

(0)                                                                                                 ,(0)                                                                                                Cancelling query.
Query was cancelled.
```
Funny, but in new version, in terminal result looks correct, like:
```
:) select sleep(1) FROM system.numbers format Values

SELECT sleep(1)
FROM system.numbers
FORMAT Values
(0)
,(0)
Cancelling query.
Query was cancelled.
```
But when I copy, result is mixed because of `\r` + spaces trick:
```
nikkochetovdev.sas.yp-c.yandex.net :) select sleep(1) + number FROM system.numbers format Values

SELECT sleep(1) + number
FROM system.numbers
FORMAT Values

(0)                                                                                                 ,(1)                                                                                                ,(2)                                                                                                ,(3)                                                                                                ,(4)                                                                                                Cancelling query.
Query was cancelled.

5 rows in set. Elapsed: 6.004 sec. 

```
#### Test 4, 5:

> Progress bar should not flicker.

Seem to be fixed

#### Test 6

> Progress bar should not flicker even after output of 10 000 rows.

Don't see difference between version. Works correct in both cases for me.


#### Test 7
Old:
```
┌─number─┐
│      4 │
└────────┘
↖ Progress: 0.00 rows, 0.00 B (0.00 rows/s., 0.00 B/s.) 
┌─number─┐
│      9 │
└────────┘
↑ Progress: 0.00 rows, 0.00 B (0.00 rows/s., 0.00 B/s.) 
┌─number─┐
│      9 │
└────────┘
↗ Progress: 0.00 rows, 0.00 B (0.00 rows/s., 0.00 B/s.) 
┌─number─┐
│      5 │
└────────┘

```

New:
```
┌─number─┐
│      9 │
└────────┘
┌─number─┐
│      9 │
└────────┘
┌─number─┐
│      7 │
└────────┘
```

New looks better for me.
Don't see other differences.

**Comment by KochetovNicolai** [comment]
TLDR: Test 3 should probably be fixed (I don't know how). It was good to copy from terminal.
Everything else works perfect.

**Comment by KochetovNicolai** [comment]
`live_view` tests are sensitive to `clickhouse-client` output. Are needed to be fixed.

**Comment by alexey-milovidov** [comment]
01396_inactive_replica_cleanup_nodes - fixed
01393_benchmark_secure_port - fixed

**Comment by KochetovNicolai** [comment]
So, now I can't copy result from client? I've expected that it will be fixed ...
